File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

factors :: Int -> [Int]
factors n = [f | f <- [1..(n `div` 2)], n `mod` f == 0]

classify :: Int -> Maybe Classification
classify n
  | n <= 0 = Nothing
  | otherwise = case compare (sum $ factors n) n of
                  EQ -> Just Perfect
                  GT -> Just Abundant
                  LT -> Just Deficient

File: clojure/perfect-numbers/src/perfect_numbers.clj
(ns perfect-numbers)

(defn factor [n]
  (loop [i (dec n)
         factors []]
    (cond
      (zero? i) factors
      (zero? (mod n i)) (recur (dec i) (cons i factors))
      :else (recur (dec i) factors))))

(defn classify [n]
  (if (neg? n)
    (throw (IllegalArgumentException. "Expected a positive integer!"))
    (let [aliquot-sum (reduce + (factor n))]
      (cond
        (= aliquot-sum n) :perfect
        (> aliquot-sum n) :abundant
        :else :deficient))))
